### PR TITLE
ENG-8833: Add support for Generic SAML IdP Integration

### PR DESCRIPTION
**Pull request by Yowgf**
## Description of the change

This PR adds support of the Generic SAML IdP integration APIs.

New data sources:

- `cyral_integration_idp_saml`

New resources:

- `cyral_integration_idp_saml_draft`
- `cyral_integration_idp_saml`

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Development

- [x] Lint rules pass locally
- [x] The code changed/added as part of this pull request has been covered with tests
- [x] All tests related to the changed code pass in development

### Code review

- [x] This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [x] Jira issue referenced in commit message and/or PR title

### Testing

- Tested with included usage examples
- ACC tests are passing
- Tested an end-to-end workflow to verify that the integration allows SSO access to control plane if properly configured.
- Check that the content returned by the data source corresponds to the reality of the API objects.
- ACC tests guarantee that data source filters work properly, but also tested manually.
- Tested corner case where SAML draft is modified but integration stays the same -- integration and draft are replaced
- Tested corner case where integration is deleted but SAML draft stays the same -- in this case, one `terraform apply` does not do the job -- Terraform state becomes inconsistent. In fact it would be impossible to make this work with a single `terraform apply`. But through introduced error handling mechanisms, a second `terraform apply` fixes the Terraform state. Note that there is a WARNING in the docs for this limitation.

### CODE REVIEW

File: cyral/resource_cyral_integration_idp_saml.go
Comment (VictorGFM) [inline cyral/resource_cyral_integration_idp_saml.go:67]: The ID is the same of the resource ID and I assume that it cannot be updated once the resource is already created, so we probably can just remove the ID and use a single structure for both Update and Read operations (See my comment above).
```suggestion
```

File: cyral/resource_cyral_integration_idp_saml.go
Comment (VictorGFM) [inline cyral/resource_cyral_integration_idp_saml.go:45]: Is this method used? I believe we can remove it.
```suggestion
```

File: cyral/model_integration_idp_generic_saml.go
Comment (VictorGFM) [inline cyral/model_integration_idp_generic_saml.go:99]: Similar to one of my comments, I believe we should include the remaining fields of the [SAMLIdentityProvider](https://github.com/cyralinc/integrations/blob/main/proto/generic-saml/generic_saml.proto#L367).

File: cyral/model_integration_idp_generic_saml.go
Comment (VictorGFM) [inline cyral/model_integration_idp_generic_saml.go:111]: I believe this should be referencing the `single_sign_on_service_url` instead, right? Also, lets include the remaining field of the [SAMLIdentityProvider](https://github.com/cyralinc/integrations/blob/main/proto/generic-saml/generic_saml.proto#L367).
```suggestion
		SingleSignOnServiceURL: d.Get("single_sign_on_service_url").(string),
```

**Review by VictorGFM — COMMENTED** [review]
Hi @Yowgf, the code looks good! Although I've provided some comments below, some of them are more of questions than suggestions. I still need to review the resources documentation, so I'll probably come up with another review for that soon.

File: cyral/resource_cyral_integration_idp_saml_draft.go
Comment (Yowgf) [reply cyral/resource_cyral_integration_idp_saml_draft.go:201]: `idp_type` is an arbitrary string that makes sense to the user as an IdP type.

File: cyral/resource_cyral_integration_idp_saml_draft.go
Comment (Yowgf) [reply cyral/resource_cyral_integration_idp_saml_draft.go:70]: It's because otherwise the response will contain the default values, and then when we write to the schema, and the user runs `terraform refresh` or `terraform apply`, the plan will not be empty. To avoid that, we can either do the above, or require that the user specify the `attributes` set, as below:

```terraform
resource "cyral_integration_idp_saml_draft" "myresource" {
    ...
    attributes{}
}
```

I have added an ACC test that captures this case, to prevent regressions.

File: cyral/resource_cyral_integration_idp_saml.go
Comment (Yowgf) [reply cyral/resource_cyral_integration_idp_saml.go:47]: The reason I chose to make them different is because the response schemas are not the same. `Update` returns `{"id":"...", "identityProvider": {...}}`, whereas `Read` returns just `{"identityProvider": {...}}`.

Having this separation will also help adapt the provider in case there is an API change in the future that is asymmetric between `Read` and `Update`.

File: cyral/resource_cyral_integration_idp_saml.go
Comment (Yowgf) [reply cyral/resource_cyral_integration_idp_saml.go:67]: I believe having this separation makes the code easier to read, and it resembles more the API, which might facilitate adaptation of the provider in case of API changes in the future (see reply to similar comment).

File: cyral/resource_cyral_integration_idp_saml.go
Comment (Yowgf) [reply cyral/resource_cyral_integration_idp_saml.go]: Done!

File: cyral/resource_cyral_integration_idp_saml.go
Comment (VictorGFM) [reply cyral/resource_cyral_integration_idp_saml.go:43]: The `Update` returns an [empty response](https://github.com/cyralinc/integrations/blob/main/proto/generic-saml/generic_saml.proto#L473), the ID of the request payload is just a path variable that is passed in the URL, so the body of the Update request and the Read response end up being the same. I mean, I'm also fine with using two different structures here, for the reasons you mentioned.

File: cyral/resource_cyral_integration_idp_saml.go
Comment (VictorGFM) [reply cyral/resource_cyral_integration_idp_saml.go:63]: I see, I think that makes sense. I would just remove the `WriteToSchema` and `ReadFromSchema` methods that are not actually being used.

File: cyral/resource_cyral_integration_idp_saml.go
Comment (Yowgf) [reply cyral/resource_cyral_integration_idp_saml.go:18]: I agree, although it makes the resource much more complex. I had chosen this just for the sake of simplicity. I am now finishing up the implementation of this option, and will ping as a reply here when I am done.

File: cyral/resource_cyral_integration_idp_saml.go
Comment (Yowgf) [reply cyral/resource_cyral_integration_idp_saml.go:91]: :facepalm: The problem is that we weren't actually using this. Now with the IdpDescription configuration option I am adding (see comment above), updates make more sense.

File: cyral/model_integration_idp_generic_saml.go
Comment (Yowgf) [reply cyral/model_integration_idp_generic_saml.go:99]: I am including the other fields in the resource as part of the addition of the `idp_description` configuration option (see comment above). Will mark as resolved once that is done.

File: cyral/model_integration_idp_generic_saml.go
Comment (Yowgf) [reply cyral/model_integration_idp_generic_saml.go]: I am including the other fields in the resource as part of the addition of the `idp_description` configuration option (see comment above). Will mark as resolved once that is done.

File: cyral/resource_cyral_integration_idp_saml.go
Comment (Yowgf) [reply cyral/resource_cyral_integration_idp_saml.go:43]: Oops sorry, I got confused there for a moment. I will remove the `ID` field in UpdateGenericSAMLRequest. Thanks! I will maintain the distinction between read and updates, though, as I really believe it makes the code easier to read.

File: cyral/resource_cyral_integration_idp_saml_draft.go
Comment (Yowgf) [reply cyral/resource_cyral_integration_idp_saml_draft.go:106]: I did try to set `id` in the past, and it didn't work. The reason it works for the S3 bucket case is because `id` is a user-settable argument (i.e. `Optional: true`). I am not sure if that makes sense in our case, since the user should have no control on the ID. What do you think?

File: cyral/resource_cyral_integration_idp_saml_draft.go
Comment (Yowgf) [reply cyral/resource_cyral_integration_idp_saml_draft.go:167]: I see. The problem I see though is that, for example in our case, we need to check if the draft still exists in the `completed` state. If we trigger a recreation every time `404` is returned, we would effectively trigger recreation every time `terraform apply` is run if there is a `cyral_integration_idp_saml` resource linked to the draft. Both the draft and the integration itself would have to be recreated. Having a fully custom function to handle the error makes more sense to me. We could construct models of common error-handling function patterns we see in the future, to avoid code duplication.

File: cyral/resource.go
Comment (Yowgf) [reply cyral/resource.go:83]: Please see my reply to similar comment above

File: cyral/resource_cyral_integration_idp_saml_draft.go
Comment (Yowgf) [reply cyral/resource_cyral_integration_idp_saml_draft.go:269]: Please see my reply to similar comment below.

**Comment by Yowgf** [comment]
@VictorGFM I was implementing the option for the user to create an integration using and IdP descriptor, but that requires adding a lot of new attributes, seems like will end up with ~500 more lines of code. I believe I would spend the whole day tomorrow just to finish this. With that said, what I will do is I will save a backup with what I already did, and push it to the next release. I will just make sure the current version is forward-compatible, so that no drastic changes will need to happen when we implement that option.

Does that sound good?

File: cyral/resource_cyral_integration_idp_saml_draft.go
Comment (VictorGFM) [inline cyral/resource_cyral_integration_idp_saml_draft.go:214]: I dont remember the conclusion we reached regarding this field, but I still think it's unnecessary and that we should consider removing it, as it adds just redundancy and makes the code more complex. Feel free to decide that though, as I dont want to block the PR with this.

**Review by VictorGFM — APPROVED** [review]
LGTM @Yowgf. I just added a single comment that I'll let you decide. I did some e2e tests with the resources and things look good, thanks for working on this, I believe this will help a lot in the implementation of future IdP modules.

File: cyral/resource_cyral_integration_idp_saml_draft.go
Comment (Yowgf) [reply cyral/resource_cyral_integration_idp_saml_draft.go]: Solved, thanks for the heads up!

**Pull request merged**
